libtenzir/builtins/functions/numeric.cpp

```diff
@@ -37,17 +38,49 @@ class round_use final : public function_use {
     // round(x, 1h) -> round to multiples of 1h
     // round(<time>, <duration>)
     // round(x, 1h) -> round so that time is multiples of 1h (for UTC timezone?)
-    auto spec = located{eval(*spec_), spec_->get_location()};
-    return round_with_spec(std::move(value), std::move(spec), ctx);
+    return round_with_spec(std::move(value), std::move(spec_.value()), ctx);
   }
 
 private:
-  static auto round_with_spec(located<series> value, located<series> spec,
+  static auto round_with_spec(located<series> value, located<duration> spec,
                               session ctx) -> series {
-    diagnostic::warning("round(_, _) is not implemented yet")
-      .primary(spec)
-      .emit(ctx);
-    return series::null(value.inner.type, value.inner.length());
+    if (spec == decltype(spec){}) {
+      diagnostic::warning("resolution must not be 0").primary(spec).emit(ctx);
+      return series::null(value.inner.type, value.inner.length());
+    }
+    auto f = detail::overload{
+      [&](const arrow::DurationArray& array) -> series {
+        auto b
+          = duration_type::make_arrow_builder(arrow::default_memory_pool());
+        check(b->Reserve(array.length()));
+        for (auto i = int64_t{0}; i < array.length(); i++) {
+          if (array.IsNull(i)) {
+            check(b->AppendNull());
+            continue;
+          }
+          const auto val = array.Value(0);
+          const auto count = spec.inner.count();
+          const auto rem = val % std::abs(count);
```

Comment:
I think you need to check for `count == 0`.

Reply:
It's done when entering the function, with a diagnostic emitted.

Reply:
Ah, right, because it's a constant. Even better!

Reply:
You could also make this an error, directly after parsing, if you want to.

Reply:
Right, just was typing the same thing, lol.

Reply:
Is there any another diagnostic you think could work better as an error too?

Reply:
> Is there any another diagnostic you think could work better as an error too?

After the execution starts, we tend to not omit errors as we generally try to keep pipelines running. So we only emit errors for catastrophic cases. However, this check here should happen during parse/compile-time and it's clearly a mistake, an error is appropriate. I don't see any other place in this function that should be promoted to an error.

But can you move the check directly before this line https://github.com/tenzir/tenzir/blob/9cc712024ce2e6a25769b8964c5bf53b00ca6967/libtenzir/builtins/functions/numeric.cpp#L152 ? Because with the way that you have written it, the error is only emitted if and when it's first called, while the other place emits it before, during function argument parsing.